### PR TITLE
Modify loadbalancer docs for external source ip preservation feature

### PR DESCRIPTION
**Pull request by girishkalele**
WIP - here is the first crack at explaining the new annotation and feature gate.

---
This change is [Reviewable](https://reviewable.io/reviews/kubernetes/kubernetes.github.io/1062)

### CODE REVIEW

File: docs/user-guide/load-balancer.md [hunk 1]
```diff
@@ -71,3 +77,61 @@ information through `kubectl`:
...
+## Loss of client source IP for external traffic
+
+Due to the implementation of this feature, the source IP for sessions as seen in the target container will *not be the original source IP* of the client. This is the default behavior as of Kubernetes v1.4. However, starting in v1.4, an optional alpha feature has been added
+that will preserve the client Source IP for GCE/GKE environments. This feature will be phased in for other cloud providers in subsequent releases.
+
+## Annotation to modify the LoadBalancer behavior for preservation of Source IP
+In 1.4, an Alpha feature has been added that changes the behavior of the external LoadBalancer feature.
+
+This feature can be activated by adding the alpha annotation to the metadata section of the Service Configuration file.
+
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: need a tick-tick-tick yaml ('''yaml, but with backicks)

Comment (thockin) [inline docs/user-guide/load-balancer.md]: err, '''json

File: docs/user-guide/load-balancer.md [hunk 2]
```diff
@@ -71,3 +77,61 @@ information through `kubectl`:
...
+        "ports": [{
+          "port": 8765,
+          "targetPort": 9376
+        }],
+        "selector": {
+          "app": "example"
+        },
+        "type": "LoadBalancer"
+      }
+    }
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: closing tickticktick

File: docs/user-guide/load-balancer.md [hunk 3]
```diff
@@ -71,3 +77,61 @@ information through `kubectl`:
...
+
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: you should explain what this does to achieve the result, including only routing locally, the fact that nodeports are not handled yet, and the basic healthchecks stuff

File: docs/user-guide/load-balancer.md [hunk 4]
```diff
@@ -37,15 +43,15 @@ Your configuration file might look like:
         "type": "LoadBalancer"
       }
     }
-
+```
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: need a blank line after tick blocks or GH may render it badly

File: docs/user-guide/load-balancer.md [hunk 5]
```diff
@@ -37,15 +43,15 @@ Your configuration file might look like:
...
 ## Using kubectl
 
 You can alternatively create the service with the `kubectl expose` command and
 its `--type=LoadBalancer` flag:
-
+```bash
     $ kubectl expose rc example --port=8765 --target-port=9376 \
         --name=example-service --type=LoadBalancer
-
+```
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: blank line after

File: docs/user-guide/load-balancer.md [hunk 6]
```diff
@@ -69,5 +75,87 @@ information through `kubectl`:
     Endpoints:    10.64.0.4:80,10.64.1.5:80,10.64.2.4:80
     Session Affinity: None
     No events.
-
+```
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: blank line

File: docs/user-guide/load-balancer.md [hunk 7]
```diff
@@ -69,5 +75,87 @@ information through `kubectl`:
...
+## Loss of client source IP for external traffic
+
+Due to the implementation of this feature, the source IP for sessions as seen in the target container will *not be the original source IP* of the client. This is the default behavior as of Kubernetes v1.4. However, starting in v1.4, an optional alpha feature has been added
+that will preserve the client Source IP for GCE/GKE environments. This feature will be phased in for other cloud providers in subsequent releases.
+
+## Annotation to modify the LoadBalancer behavior for preservation of Source IP
+In 1.4, an Alpha feature has been added that changes the behavior of the external LoadBalancer feature.
+
+This feature can be activated by adding the alpha annotation to the metadata section of the Service Configuration file.
+```json
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: blank line before

File: docs/user-guide/load-balancer.md [hunk 8]
```diff
@@ -69,5 +75,87 @@ information through `kubectl`:
...
+        "type": "LoadBalancer"
+      }
+    }
+```
+
+### Alpha Feature Gate for the 'service.alpha.kubernetes.io/external-traffic' annotation
+
+Alpha features are not enabled by default, they must be enabled using the release gate command line flags
+for kube-controller-manager and kube-proxy.
+See TBD for more details on feature gate flags.
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: TBD

File: docs/user-guide/load-balancer.md [hunk 9]
```diff
@@ -69,5 +75,87 @@ information through `kubectl`:
...
+        }],
+        "selector": {
+          "app": "example"
+        },
+        "type": "LoadBalancer"
+      }
+    }
+```
+
+### Alpha Feature Gate for the 'service.alpha.kubernetes.io/external-traffic' annotation
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: also discuss that this does not work on all cloud providers yet

File: docs/user-guide/load-balancer.md [hunk 10]
```diff
@@ -69,5 +75,87 @@ information through `kubectl`:
...
+distribution will be seen, even without weights.
+
+Once the external load balancers provide weights, this functionality can be added to the LB programming path.
+*Future Work: No support for weights is provided for the 1.4 release, but may be added at a future date*
+
+Internal pod to pod traffic should behave similar to ClusterIP services, with equal probability across all pods.
+
+## How is this feature implemented ?
+
+### GCE/GKE Equal Load Distribution (Source IP Not Preserved)
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: This doc is not about GCE.  It should be totally platform neutral

File: docs/user-guide/load-balancer.md (hunk 8 above)
Comment (girishkalele) [inline docs/user-guide/load-balancer.md]: Added link to Jeff's proposal - there are no kubernetes.io docs I could find.

File: docs/user-guide/load-balancer.md (hunk 9 above)
Comment (girishkalele) [inline docs/user-guide/load-balancer.md]: done

File: docs/user-guide/load-balancer.md (hunk 10 above)
Comment (girishkalele) [inline docs/user-guide/load-balancer.md]: Removed GCE/GKE section.

File: docs/user-guide/load-balancer.md [hunk 11]
```diff
@@ -69,5 +82,67 @@ information through `kubectl`:
...
+Alpha features are not enabled by default, they must be enabled using the release gate command line flags
+for kube-controller-manager and kube-proxy.
+See [https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/runtimeconfig.md](Runtime feature flags proposal) for more details on feature gate flags.
+
+If this feature is not enabled in your cluster, this service configuration will be rejected.
+
+Also note this feature is not currently implemented for all cloudproviders/environments.
+
+### Caveats and Limitations when preserving source IPs
+
```
Comment (thockin) [inline docs/user-guide/load-balancer.md]: Still want a note that this does not yet work on nodePorts, so "proxy style" balancers can not use this yet.

**Comment by thockin** [comment]
One nit then docs LGTM

**Comment by girishkalele** [comment]
Fixed nit...marking LGTM.

**Comment by girishkalele** [comment]
Rebased.

**Comment by girishkalele** [comment]
Fixes #1058

**Comment by thockin** [comment]
needs a rebase

**Comment by girishkalele** [comment]
I tried the "Update branch" button but it does a pull instead of a rebase.

Rebased manually instead.

**Comment by girishkalele** [comment]
cc @kubernetes/docs

**Pull request merged**